homeassistant/components/tplink_omada/update.py

```diff
@@ -0,0 +1,149 @@
+        )
+        for d in devices
+    ]
+
+
+async def _poll_firmware_updates(
+    client: OmadaSiteClient,
+) -> dict[str, FirmwareUpdateStatus]:
+    """Poll the state of Omada Devices firmware update availability."""
+    return {d.device.mac: d for d in await _get_firmware_updates(client)}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switches."""
+    controller: OmadaSiteController = hass.data[DOMAIN][config_entry.entry_id]
+    omada_client = controller.omada_client
+
+    devices = await omada_client.get_devices()
+
+    coordinator = OmadaCoordinator[FirmwareUpdateStatus](
+        hass,
+        omada_client,
+        "Firmware Updates",
+        _poll_firmware_updates,
+        poll_delay=6 * 60 * 60,
+    )
+
+    entities: list = []
+    for device in devices:
+        entities.append(OmadaDeviceUpdate(coordinator, device))
+
+    async_add_entities(entities)
+    await coordinator.async_request_refresh()
+
+
+class OmadaDeviceUpdate(
+    OmadaDeviceEntity[FirmwareUpdateStatus],
+    UpdateEntity,
+):
+    """Firmware update status for Omada SDN devices."""
+
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL
+        | UpdateEntityFeature.PROGRESS
+        | UpdateEntityFeature.RELEASE_NOTES
+    )
+    _firmware_update: OmadaFirmwareUpdate = None
+
+    def __init__(
+        self,
+        coordinator: OmadaCoordinator[FirmwareUpdateStatus],
+        device: OmadaListDevice,
+    ) -> None:
+        """Initialize the update entity."""
+        super().__init__(coordinator, device)
+
+        self._mac = device.mac
+        self._device = device
+        self._omada_client = coordinator.omada_client
+
+        self._attr_unique_id = f"{device.mac}_firmware"
+        self._attr_has_entity_name = True
+        self._attr_name = "Firmware Update"
+        self._refresh_state()
+
+    def _refresh_state(self) -> None:
+        if self._firmware_update and self._device.need_upgrade:
+            self._attr_installed_version = self._firmware_update.current_version
+            self._attr_latest_version = self._firmware_update.latest_version
+        else:
+            self._attr_installed_version = self._device.firmware_version
+            self._attr_latest_version = self._device.firmware_version
+        self._attr_in_progress = self._device.fw_download
+
+        if self._attr_in_progress:
+            # While firmware update is in progress, poll more frequently
+            async_call_later(self.hass, 60, self._request_refresh)
+
+    async def _request_refresh(self, _now: Any) -> None:
+        await self.coordinator.async_request_refresh()
+
+    def release_notes(self) -> str | None:
+        """Get the release notes for the latest update."""
+        if self._firmware_update:
+            return str(self._firmware_update.release_notes)
```

Comment:
Why is this being converted to a string? Isn't it a string yet? Or is it a typing issue?
In case of the latter, the string wrapping seems to be unneeded.

Reply:
> error: Returning Any from function declared to return "Optional[str]"  [no-any-return]

It means upstream typing isn't correct (in the library). Either fix that, or cast instead of convert.

Reply:
Yeah, I thought str _was_ a cast :)
The library does claim it's a str property, and VS Code agrees, so I'm not sure what's actually wrong.
<img width="365" alt="image" src="https://user-images.githubusercontent.com/10632972/224838017-14f8e606-1cb3-4f26-aefc-e803773787c0.png">

```
    @property
    def release_notes(self) -> str:
        """Release notes for the new firmware."""
        return self._data["fwReleaseLog"]
```
Perhaps that it's implicitly converting from `Any` in the return?

Reply:
The library doesn't comply with PEP-561, the `py.typed` file is missing. This results in everything being `Any`.

<https://peps.python.org/pep-0561/>

Reply:
Ok, looks like I have more stuff to read. I'll get this straightened out.
Thanks, and sorry for my ignorance.

Reply:
Don't worry! It is simple to fix/do and happy to help as well.

Just add an empty  `py.typed` file in your package, here is an example from one of my own:

<https://github.com/frenck/python-elgato/tree/main/src/elgato>

It indicates your package has been typed.

../Frenck